File: iq/10/chain/find.go
// Copyright 2026 Samvel Khalatyan. All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package chain

import "iter"

func Find(nn []int) []int {
	set := make(map[int]struct{})
	for _, n := range nn {
		set[n] = struct{}{}
	}
	var longestChain []int
	for chain := range findChains(set) {
		if len(chain) <= len(longestChain) {
			continue
		}
		longestChain = chain[:]
	}
	return longestChain
}

type C []int

func findChains(nn map[int]struct{}) iter.Seq[C] {
	return func(yield func(C) bool) {
		for n := range nn {
			if _, ok := nn[n-1]; ok {
				continue
			}
			chain := newChain(nn, n)
			if !yield(chain) {
				break
			}
		}
	}
}

func newChain(nn map[int]struct{}, start int) C {
	cc := []int{start}
	exists := func(n int) bool {
		_, ok := nn[n]
		return ok
	}
	for n := start + 1; exists(n); n++ {
		cc = append(cc, n)
	}
	return cc
}

File: iq/10/chain/find.cc
// Copyright 2026 Samvel Khalatyan. All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

//
// NOLINTNEXTLINE
//go:build ignore

#include "iq/10/chain/find.h"

#include <unordered_set>
#include <vector>

namespace iq::chain {
namespace {

std::vector<int> makeChain(const std::unordered_set<int>& nn, int n) {
  std::vector<int> chain = {n};
  for (++n; nn.contains(n); ++n) {
    chain.push_back(n);
  }
  return chain;
}

}  // namespace

std::vector<int> Find(const std::vector<int>& nn) {
  if (nn.empty()) {
    return {};
  }
  std::unordered_set<int> seen{nn.begin(), nn.end()};
  std::vector<int> longest_chain;
  for (const int n : seen) {
    if (const int p = n - 1; seen.contains(p)) {
      continue;
    }
    if (std::vector<int> chain = makeChain(seen, n);
        chain.size() > longest_chain.size()) {
      longest_chain = std::move(chain);
    }
  }
  return longest_chain;
}

}  // namespace iq::chain
